File: api/src/common/policy/opa.service.ts
import { Injectable, Logger } from "@nestjs/common";
import { ConfigService } from "@nestjs/config";

interface OpaAllowResponse {
  result?: boolean;
}

@Injectable()
export class OpaService {
  private readonly logger = new Logger(OpaService.name);
  private readonly opaUrl: string;

  constructor(configService: ConfigService) {
    this.opaUrl = configService.get<string>("OPA_URL", "http://localhost:8181");
  }

  async canViewPii(input: Record<string, unknown>): Promise<boolean> {
    const url = `${this.opaUrl}/v1/data/illamhelp/pii/allow`;

    try {
      const response = await fetch(url, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ input })
      });

      if (!response.ok) {
        this.logger.warn(`OPA returned non-200 status: ${response.status}`);
        return false;
      }

      const payload = (await response.json()) as OpaAllowResponse;
      return payload.result === true;
    } catch (error) {
      this.logger.warn(`OPA request failed: ${(error as Error).message}`);
      return false;
    }
  }
}

File: api-java/src/main/java/com/illamhelp/api/consent/OpaService.java
package com.illamhelp.api.consent;

import com.illamhelp.api.config.AppProperties;
import java.util.Map;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;

@Service
public class OpaService {
  private final AppProperties properties;
  private final RestClient restClient;

  public OpaService(AppProperties properties, RestClient.Builder builder) {
    this.properties = properties;
    this.restClient = builder.build();
  }

  public boolean canViewPii(Map<String, Object> input) {
    try {
      Map<?, ?> response = restClient.post()
          .uri(properties.opaUrl() + "/v1/data/illamhelp/pii/allow")
          .body(Map.of("input", input))
          .retrieve()
          .body(Map.class);
      return Boolean.TRUE.equals(response == null ? null : response.get("result"));
    } catch (RuntimeException exception) {
      return false;
    }
  }
}
